pkg/controller/elasticquota.go

```diff
@@ -0,0 +1,348 @@
+	used := newZeroUsed(eq)
+	pods, err := ctrl.podLister.Pods(namespace).List(labels.NewSelector())
+	if err != nil {
+		return nil, err
+	}
+	for _, p := range pods {
+		if p.Status.Phase == v1.PodRunning {
+			used = quota.Add(used, computePodResourceRequest(p))
+		}
+	}
+	return used, nil
+}
+
+// eqAdded reacts to a ElasticQuota creation
+func (ctrl *ElasticQuotaController) eqAdded(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	klog.V(5).Infof("Enqueue new elastic quota key %v", key)
+	ctrl.eqQueue.AddRateLimited(key)
+}
+
+// eqUpdated reacts to a ElasticQuota update
+func (ctrl *ElasticQuotaController) eqUpdated(oldObj, newObj interface{}) {
+	ctrl.eqAdded(newObj)
+}
+
+// eqDeleted reacts to a ElasticQuota deletion
+func (ctrl *ElasticQuotaController) eqDeleted(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	klog.V(5).Infof("Enqueue deleted elastic quota key %v", key)
+	ctrl.eqQueue.AddRateLimited(key)
+}
+
+// podAdded reacts to a pod creation
+func (ctrl *ElasticQuotaController) podAdded(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	list, err := ctrl.eqLister.ElasticQuotas(pod.Namespace).List(labels.Everything())
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	if len(list) == 0 {
+		return
+	}
+	// todo(yuchen-sun) When elastic quota supports multiple instances in a namespace, modify this
+	ctrl.eqAdded(list[0])
+}
+
+// podUpdated reacts to a pod update
+func (ctrl *ElasticQuotaController) podUpdated(oldObj, newObj interface{}) {
+	oldPod := oldObj.(*v1.Pod)
+	newPod := newObj.(*v1.Pod)
+	if newPod.ResourceVersion == oldPod.ResourceVersion {
+		return
+	}
+	ctrl.podAdded(newObj)
+}
+
+// podDeleted reacts to a pod delete
+func (ctrl *ElasticQuotaController) podDeleted(obj interface{}) {
+	_, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	ctrl.podAdded(obj)
+}
+
+// computePodResourceRequest returns a v1.ResourceList that covers the largest
+// width in each resource dimension. Because init-containers run sequentially, we collect
+// the max in each dimension iteratively. In contrast, we sum the resource vectors for
+// regular containers since they run simultaneously.
+//
+// If Pod Overhead is specified and the feature gate is set, the resources defined for Overhead
+// are added to the calculated Resource request sum
+//
+// Example:
+//
+// Pod:
+//   InitContainers
+//     IC1:
+//       CPU: 2
+//       Memory: 1G
+//     IC2:
+//       CPU: 2
+//       Memory: 3G
+//   Containers
+//     C1:
+//       CPU: 2
+//       Memory: 1G
+//     C2:
+//       CPU: 1
+//       Memory: 1G
+//
+// Result: CPU: 3, Memory: 3G
+func computePodResourceRequest(pod *v1.Pod) v1.ResourceList {
```

Comment:
I assume this logic is copied from somewhere? Can we reuse it rather than copying here?

Reply:
This logic duplicates the Capacity scheduling logic, but  it's not exactly the same 
I think we can modify the logic in the capacity scheduling so that it doesn't have to maintain additional information about the eq usage.

Reply:
I think we can use another PR to get rid of the duplicate logic in Capacity scheduling

Reply:
SG.